src/org/rascalmpl/library/lang/rascal/tests/concrete/recovery/ParseErrorDisambiguator.java

```diff
@@ -0,0 +1,174 @@
+        if (ProductionAdapter.isSkipped(appl.getProduction())) {
+            return new ScoredTree(appl, ((IList) appl.get(1)).length());
+        }
+
+        IList args = TreeAdapter.getArgs(appl);
+        int totalScore = 0;
+        IValue[] disambiguatedArgs = new IValue[args.size()];
+
+        int index=0;
+        boolean anyChanges = false;
+
+        // Disambiguate and score all children
+        for (IValue arg : args) {
+            ScoredTree disambiguatedArg = disambiguate((IConstructor) arg);
+            totalScore += disambiguatedArg.score;
+            disambiguatedArgs[index++] = disambiguatedArg.tree;
+            if (disambiguatedArg.tree != arg) {
+                anyChanges = true;
+            }
+        }
+
+        // Only build a new tree if at least one of the arguments has changed
+        ITree resultTree;
+        if (anyChanges) {
+            IListWriter argWriter = rascalValues.listWriter();
+            for (IValue arg : disambiguatedArgs) {
+                argWriter.append(arg);
+            }
+            resultTree = TreeAdapter.setArgs(appl, argWriter.done());
+        } else {
+            resultTree = appl;
+        }
+
+
+        return new ScoredTree(resultTree, totalScore);
+    }
+
+    private ScoredTree disambiguateAmb(ITree amb) {
+        ISet alts = (ISet) amb.get(0);
+
+        ISetWriter nonErrorAlts = null;
+        ScoredTree bestErrorAlt = null;
+        for (IValue alt : alts) {
+            ScoredTree disambiguatedAlt = disambiguate((IConstructor) alt);
+            if (disambiguatedAlt.score == 0) {
+                // Non-error tree
+                if (nonErrorAlts == null) {
+                    nonErrorAlts = rascalValues.setWriter();
+                }
+                nonErrorAlts.insert(disambiguatedAlt.tree);
+            } else {
+                // Only keep the best of the error trees
+                if (bestErrorAlt == null || bestErrorAlt.score > disambiguatedAlt.score) {
+                    bestErrorAlt = disambiguatedAlt;
+                }
+            }
+        }
+
+        if (nonErrorAlts == null) {
+            return bestErrorAlt;
+        }
+        
+        ISet newAlts = nonErrorAlts.done();
+        int altCount = newAlts.size();
+
+        ITree resultTree;
+        if (altCount == alts.size()) {
+            // All children are without errors, return the original tree
+            resultTree = amb;
+        } else if (altCount == 1) {
+            // One child without errors remains, dissolve the amb tree
+            resultTree = (ITree) newAlts.iterator().next();
+        } else {
+            // Create a new amb tree with the remaining non-error trees
+            resultTree = rascalValues.amb(newAlts);
+
+            // We have an ambiguity between non-error trees
+            if (!allowAmbiguity) {
+                throw new Ambiguous(resultTree);
+            }
+        }
+
+        return new ScoredTree(resultTree, 0);
```

Comment:
why is the value 0? we can still have errors in the amb trees right? the original rascal code would take into account all nested error scores as well.